python-checks/src/main/java/org/sonar/python/checks/DictionaryDuplicateKeyCheck.java

```diff
@@ -0,0 +1,132 @@
+ */
+package org.sonar.python.checks;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.tree.DictionaryLiteral;
+import org.sonar.plugins.python.api.tree.DictionaryLiteralElement;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.KeyValuePair;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.NumericLiteral;
+import org.sonar.plugins.python.api.tree.StringElement;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+
+@Rule(key = "S5780")
+public class DictionaryDuplicateKeyCheck extends PythonSubscriptionCheck {
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.DICTIONARY_LITERAL, ctx -> {
+      DictionaryLiteral dictionaryLiteral = (DictionaryLiteral) ctx.syntaxNode();
+      Set<Integer> issueIndexes = new HashSet<>();
+      for (int i = 0; i < dictionaryLiteral.elements().size(); i++) {
+        if (!dictionaryLiteral.elements().get(i).is(Tree.Kind.KEY_VALUE_PAIR) || issueIndexes.contains(i)) {
+          continue;
+        }
+        Expression key = ((KeyValuePair) dictionaryLiteral.elements().get(i)).key();
+        List<Tree> duplicateKeys = findIdenticalKeys(i, dictionaryLiteral.elements(), issueIndexes);
+        if (!duplicateKeys.isEmpty()) {
+          PreciseIssue issue = ctx.addIssue(key, "Change or remove duplicates of this key.");
+          duplicateKeys.forEach(d -> issue.secondary(d, "Duplicate key"));
+        }
+      }
+    });
+  }
+
+  private List<Tree> findIdenticalKeys(int startIndex, List<DictionaryLiteralElement> elements, Set<Integer> issueIndexes) {
+    Expression key = ((KeyValuePair) elements.get(startIndex)).key();
+    List<Tree> duplicates = new ArrayList<>();
+    for (int i = startIndex + 1; i < elements.size() ; i++) {
+      if (!elements.get(i).is(Tree.Kind.KEY_VALUE_PAIR)) {
+        continue;
+      }
+      Expression comparedKey = ((KeyValuePair) elements.get(i)).key();
+      if (isSameKey(key, comparedKey)) {
+        issueIndexes.add(i);
+        duplicates.add(comparedKey);
+      }
+    }
+    return duplicates;
+  }
+
+  private boolean isSameKey(Expression key, Expression comparedKey) {
+    if (key.is(Tree.Kind.STRING_LITERAL) && comparedKey.is(Tree.Kind.STRING_LITERAL)) {
+      return areEquivalentStringLiterals((StringLiteral) key, (StringLiteral) comparedKey);
+    }
+    if (isANumber(key) && isANumber(comparedKey)) {
+      return areEquivalentNumbers(key, comparedKey);
+    }
+    return !key.is(Tree.Kind.CALL_EXPR) && CheckUtils.areEquivalent(key, comparedKey);
```

Comment:
1. On the CALL_EXPR: excluding only the `CALL_EXPR` might lead to false positives. Consider the following example:

```
class Foo:
      def __init__(self):
        self.counter = 0
      def __getitem__(self, i):
        self.counter = self.counter + 1
        return i + self.counter

    foo = Foo()
    { foo[0]: 'a', foo[0]: 'b' } # that's ok, creates 2-entry dictionary
```
This gives a false positive, because it thinks that `foo[0]` results in the same key every time.
To avoid this, one would have to exclude all constructs that might have side-effects
(randomness, io, network, mutable state etc.) - it's not only function calls. I'm not sure whether it's feasible. Maybe it would be safer to restrict it to fewer cases that are explicitly allowed, instead of excluding CALL_EXPR?

------------

2. On the `CheckUtils.areEquivalent`:

  `{ (0o10, 'a'): 1, (8, "a"): 2 }` gives a false negative. That a pity, because
  there already are much nicer `areEquivalentNumbers` and `areEquivalentStringLiterals`, but they
  are not used by `CheckUtils.areEquivalent`. Would it maybe be possible to implement a recursive
  variant of `CheckUtils.areEquivalent` that works with `areEquivalent{Numbers,StringLiterals}`
  at the leaves?

Reply:
On the first case, that would absolutely be a FP which I'm not sure is worth addressing (I believe it depends on how common the approach is, which I must admit I have no idea).

One thing to point out is that a similar FP would happen on a member access `a.b` which uses a descriptor that somehow doesn't return the same value at each access. We discussed this case with @nicolas-harraudeau-sonarsource  and decided to accept the risk of FP.

The problem with such cases is that we don't always have a strong way of knowing whether the default behavior is being overridden or not, so it would be either accepting FPs for the odd case, or accepting FNs for the most common cases. I'd argue we see the result on Peach for this one to see if there are any further restrictions we should make. WDYT @nicolas-harraudeau-sonarsource ?

For the second case, that's a good catch and it made me realize the current implementation also reports FPs when two identical tuples contain function calls. This is fixed in af74d2bb08b0c29d5d5b4ec9ca09a5b399d3dc90.

Reply:
1. I have to admit that the example is completely synthetic. I agree, it's probably very uncommon.

2. LGTM